### ISSUE TITLE
Templates defined in Dynamic Mapping are NOT processed in order

### ISSUE THREAD

**Issue by androidkencai** [elastic/elasticsearch#18625]
**Elasticsearch version**:
1.7 or 2.2 

**JVM version**:
1.8.0_66

**OS version**:
OSX 10.11.5

**Description of the problem including expected versus actual behavior**:
According to the [dynamic-templates](https://www.elastic.co/guide/en/elasticsearch/reference/current/dynamic-templates.html)
> Templates are processed in order — the first matching template wins.

in the test case below, it's expected to hit only the first template definition, **path_match_str_def**, but it seemed hit the one next and got the exceptions attached. 

**Steps to reproduce**:
 1.mapping template 
```javascript  
"mappings": {
    "sdr": {
      "dynamic_templates": [
        {
          "path_match_str_def": {
            "path_match": "mypath.*",
            "match_mapping_type": "string",
            "mapping": {
              "type": "string",
              "index": "not_analyzed",
              "doc_values": true
            }
          }
        },
        {
          "path_match_str_int": {
            "path_match": "mypath.var_i*",
            "mapping": {
              "type": "long",
              "index": "not_analyzed",
              "doc_values": true
            }
          }
        }
      ],
      "properties": {
        "mypath": {
          "type": "object",
          "properties": {
            "root": {
              "type": "string",
              "index": "not_analyzed",
              "doc_values": true
            }
          }
        }
      }
    }
  }
```
 2.document
`{
  "mypath": {
    "var_istring": "abcdef"
  }
}`
 3.got exceptions below

```java
RemoteTransportException[[Living Laser][127.0.0.1:9300][indices:data/write/index[p]]]; nested: MapperParsingException[failed to parse [mypath.var_istring]]; nested: NumberFormatException[For input string: "abcdef"];
Caused by: MapperParsingException[failed to parse [mypath.var_istring]]; nested: NumberFormatException[For input string: "abcdef"];
	at org.elasticsearch.index.mapper.FieldMapper.parse(FieldMapper.java:343)
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:318)
	at org.elasticsearch.index.mapper.DocumentParser.parseAndMergeUpdate(DocumentParser.java:765)
	at org.elasticsearch.index.mapper.DocumentParser.parseDynamicValue(DocumentParser.java:652)
	at org.elasticsearch.index.mapper.DocumentParser.parseValue(DocumentParser.java:451)
	at org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:271)
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:315)
	at org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:335)
	at org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:261)
	at org.elasticsearch.index.mapper.DocumentParser.innerParseDocument(DocumentParser.java:131)
	at org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:79)
	at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:304)
	at org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:547)
	at org.elasticsearch.index.shard.IndexShard.prepareIndexOnPrimary(IndexShard.java:529)
	at org.elasticsearch.action.index.TransportIndexAction.prepareIndexOperationOnPrimary(TransportIndexAction.java:211)
	at org.elasticsearch.action.index.TransportIndexAction.executeIndexRequestOnPrimary(TransportIndexAction.java:223)
	at org.elasticsearch.action.index.TransportIndexAction.shardOperationOnPrimary(TransportIndexAction.java:157)
	at org.elasticsearch.action.index.TransportIndexAction.shardOperationOnPrimary(TransportIndexAction.java:65)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase.doRun(TransportReplicationAction.java:595)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:263)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:260)
	at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:350)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NumberFormatException: For input string: "abcdef"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Long.parseLong(Long.java:589)
	at java.lang.Long.parseLong(Long.java:631)
	at org.elasticsearch.common.xcontent.support.AbstractXContentParser.longValue(AbstractXContentParser.java:145)
	at org.elasticsearch.index.mapper.core.LongFieldMapper.innerParseCreateField(LongFieldMapper.java:275)
	at org.elasticsearch.index.mapper.core.NumberFieldMapper.parseCreateField(NumberFieldMapper.java:241)
	at org.elasticsearch.index.mapper.FieldMapper.parse(FieldMapper.java:335)
	... 26 more
 ```

**Comment by jpountz** [comment]
I can reproduce on master too. The issue is due to the fact that in some cases elasticsearch will apply templates that do not define a `match_mapping_type` before those that do.

**Comment by jpountz** [comment]
I am surprised nobody found it before as this bug seems to be almost 5 years old!

**Comment by jpountz** [comment]
Thanks @androidkencai for reporting this bug. It will be addressed in Elasticsearch 2.4.0 and 5.0.0.

**Comment by androidkencai** [comment]
@jpountz you're welcome. We're still on Kibana3. Will it be addressed in 1.7.x?

**Comment by s1monw** [comment]
@androidkencai I am almost certain that there won't be another `1.7.x` release. you should look into upgradeing
